### PR TITLE
log_sigmoid: Use log1p for improved precision

### LINKED ISSUE #20972: logsigmoid fails on -inf
## 🐛 Bug

torch.nn.functional.logsigmoid returns nan when its input is -inf, the correct return value should be -inf. Algorithm should be adjusted to return the correct value (and correct derivative), or the special case should be tested for and replaced.

The impact of the issue is that if a -inf is present in an input tensor the output is corrupted and the gradient will fail to compute for the full tensor, even though -inf should not be problematic.

In fixing the error the gradient should also be checked and should be 1.0 (currently gradient calculation fails due to nan.)

## To Reproduce

1. Run the code snippet below and get a nan result.
```python
>>> x = torch.tensor(-float('Inf'),requires_grad=True)
>>> y = torch.nn.functional.logsigmoid(x)
>>> y.backward()
>>> print('y = '+str(y.item()))
y = nan
>>> print('dy/dx = '+str(x.grad.item()))
dy/dx = nan
```

## Expected behavior
```python
>>> x = torch.tensor(-float('Inf'),requires_grad=True)
>>> y = torch.nn.functional.logsigmoid(x)
>>> y.backward()
>>> print('y = '+str(y.item()))
y = -inf
>>> print('dy/dx = '+str(x.grad.item()))
dy/dx = 1.0
```

## Environment
 - PyTorch Version (e.g., 1.0): 1.1 (cpu only)
 - OS (e.g., Linux): Windows
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source): n/a
 - Python version:3.7.1
 - CUDA/cuDNN version: n/a
 - GPU models and configuration: n/a
 - Any other relevant information: n/a

## Additional context

Problem should occur generally as it is a flaw in the algorithm, I have not tested on a gpu, only on a cpu.

### PR DESCRIPTION
**Pull request by peterbell10**
Fixes #20972

log_sigmoid calculates something like `log(1 + x)` where x is always a
positive number less than one. This wastes floating point precision
because the exponent always becomes zero. Instead, using
`log1p(x)` gives the full mantissa precision around `x=0`.

This also fixes infinity propagation because the old code does,
`exp(in - in)` when `in` is negative. Which for infinity, results in a
NaN instead of 0.

cc @albanD @mruberry @jbschlosser @walterddr

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
I'm not sure about the new gradient formula?

Do you have any numbers on the perf implication?

**Review by albanD — APPROVED** [review]
SGTM
